confmap/confmap.go

```diff
@@ -309,12 +309,14 @@ func isStringyStructure(t reflect.Type) bool {
 // When a value has been loaded from an external source via a provider, we keep both the
 // parsed value and the original string value. This allows us to expand the value to its
 // original string representation when decoding into a string field, and use the original otherwise.
-func useExpandValue() mapstructure.DecodeHookFuncType {
+func useExpandValue() mapstructure.DecodeHookFuncValue {
 	return func(
-		_ reflect.Type,
-		to reflect.Type,
-		data any,
+		fromVal reflect.Value,
+		toVal reflect.Value,
 	) (any, error) {
+		to := toVal.Type()
```

Comment:
This will run into the same issue https://github.com/open-telemetry/opentelemetry-collector/pull/12996#issuecomment-2862859367. I don't remember now, but there is safe way in reflect to check if `toVal.Type()` would actually succeed, and if not this hook should just return.

Reply:
Okay, I will work on that. I just realized that https://github.com/go-viper/mapstructure/blob/c29fc28e7927f11614d6f7e0fbf108a5c75750bb/decode_hooks.go#L95 is also a source of panics, so I may have to rewrite the compose function as well

Reply:
This is hard stuff. Thanks @mx-psi.

Reply:
So it looks like calling `IsValid` is the thing we need. I ended up moving this logic into the compose hook.

The changes in this patch are no-op as of now but would make the test I mentioned in #12996 pass without panicking. PTAL!